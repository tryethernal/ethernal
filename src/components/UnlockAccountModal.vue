<template>
<v-dialog v-model="dialog" max-width="430">
    <v-card>
        <v-card-title class="headline">Unlock Account</v-card-title>

        <v-card-text>
            <v-alert type="error" v-if="errorMessage">{{ errorMessage }}</v-alert>
            <v-alert type="success" v-if="successMessage">{{ successMessage }}</v-alert>
            <div>
                Set private key for <b>{{ options.address }}</b> in order to use it for methods call. If you've already set one in the past, it will override it.
            </div>
            <div>
                Private keys are encrypted server side with AES 256 CBC, and stored encrypted. We strongly recommend to not use accounts with any value.
            </div>
            <v-text-field id="privateKey" outlined class="mt-2" v-model="privateKey" label="Key*" required></v-text-field>
        </v-card-text>

        <v-card-actions>
            <v-spacer></v-spacer>
            <v-btn color="primary" text @click="close()">Close</v-btn>
            <v-btn id="unlockAccount" color="primary" :loading="loading" :disabled="!privateKey" @click="unlockAccount(options.address, privateKey)">Unlock</v-btn>
        </v-card-actions>
    </v-card>
</v-dialog>
</template>
<script>
const ethers = require('ethers');

export default {
    name: 'UnlockAccountModal',
    data: () => ({
        dialog: false,
        resolve: null,
        reject: null,
        options: {},
        privateKey: null,
        loading: false,
        errorMessage: null,
        successMessage: null
    }),
    methods: {
        open: function(options) {
            this.dialog = true;
            this.options = options ;
            return new Promise((resolve, reject) => {
                this.resolve = resolve;
                this.reject = reject;
            })
        },
        close: function() {
            this.resolve(false);
            this.reset();
        },
        unlockAccount: function(address, privateKey) {
            try {
                this.loading = true;
                this.errorMessage = null;
                this.successMessage = null;

                const wallet = new ethers.Wallet(privateKey);

                if (wallet.address.toLowerCase() != address)
                    throw { code: 'WALLET_MISMATCH'};

                this.server.storeAccountPrivateKey(address, privateKey)
                    .then(() => {
                        this.successMessage = 'Account unlocked.';
                    })
                    .catch(console.log)
                    .finally(() => this.loading = false);
                } catch(error) {
                    this.loading = false;
                    switch (error.code) {
                        case 'WALLET_MISMATCH':
                            this.errorMessage = `Private key doesn't match the address.`;
                            break;
                        case 'INVALID_ARGUMENT':
                        default:
                            this.errorMessage = 'Invalid private key.'

                    }
                }
        },
        reset: function() {
            this.dialog = false;
            this.resolve = null;
            this.reject = null;
            this.options = {};
            this.pkey = null;
        }
    }
}
</script>
