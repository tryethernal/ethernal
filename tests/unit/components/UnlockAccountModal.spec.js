jest.mock('ethers', () => {
    return {
        Wallet: jest.fn(() => { return { address: '0x123' }})
    };
});
import ethers from 'ethers';
import MockHelper from '../MockHelper';

import UnlockAccountModal from '@/components/UnlockAccountModal.vue';

describe('UnlockAccountModal.vue', () => {
    let helper;

    beforeEach(async () => {
        helper = new MockHelper();        
    });

    it('Should let you unlock with a private key', async () => {
        jest.spyOn(ethers, 'Wallet').mockImplementation(() => { return { address: '0x123' }});
        const wrapper = helper.mountFn(UnlockAccountModal);

        await wrapper.setData({ options: { address: '0x123' }, dialog: true });
        await wrapper.find('#privateKey').setValue('0x456');
        await wrapper.find('#unlockAccount').trigger('click');
        await new Promise(process.nextTick);

        expect(wrapper.vm.successMessage).toBe('Account unlocked.')
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display a warning if address mismatch', async () => {
        jest.spyOn(ethers, 'Wallet').mockImplementation(() => { return { address: '0xabc' }});
        const wrapper = helper.mountFn(UnlockAccountModal);

        await wrapper.setData({ options: { address: '0x123' }, dialog: true });
        await wrapper.find('#privateKey').setValue('0x456');
        await wrapper.find('#unlockAccount').trigger('click');
        await new Promise(process.nextTick);

        expect(wrapper.vm.errorMessage).toBe(`Private key doesn't match the address.`);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display a warning if private key format is incorrect', async () => {
        jest.spyOn(ethers, 'Wallet').mockImplementation(() => { throw { code: 'INVALID_ARGUMENT' }});
        const wrapper = helper.mountFn(UnlockAccountModal);

        await wrapper.setData({ options: { address: '0x123' }, dialog: true });
        await wrapper.find('#privateKey').setValue('ethernal');
        await wrapper.find('#unlockAccount').trigger('click');
        await new Promise(process.nextTick);

        expect(wrapper.vm.errorMessage).toBe(`Invalid private key.`);
        expect(wrapper.html()).toMatchSnapshot();
    });    
});
