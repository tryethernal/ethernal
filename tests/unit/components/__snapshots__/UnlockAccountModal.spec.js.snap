// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnlockAccountModal.vue Should display a warning if address mismatch 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 430px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Unlock Account</div>
          <div class="v-card__text">
            <div role="alert" class="v-alert v-sheet theme--dark error">
              <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark"></i>
                <div class="v-alert__content">Private key doesn't match the address.</div>
              </div>
            </div>
            <!---->
            <div>
              Set private key for <b>0x123</b> in order to use it for methods call. If you've already set one in the past, it will override it.
            </div>
            <div>
              Private keys are encrypted server side with AES 256 CBC, and stored encrypted. We strongly recommend to not use accounts with any value.
            </div>
            <div class="v-input mt-2 v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <fieldset aria-hidden="true">
                    <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                  </fieldset>
                  <div class="v-text-field__slot"><label for="privateKey" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Key*</label><input required="required" id="privateKey" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary" id="unlockAccount"><span class="v-btn__content">Unlock</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`UnlockAccountModal.vue Should display a warning if private key format is incorrect 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 430px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Unlock Account</div>
          <div class="v-card__text">
            <div role="alert" class="v-alert v-sheet theme--dark error">
              <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark"></i>
                <div class="v-alert__content">Invalid private key.</div>
              </div>
            </div>
            <!---->
            <div>
              Set private key for <b>0x123</b> in order to use it for methods call. If you've already set one in the past, it will override it.
            </div>
            <div>
              Private keys are encrypted server side with AES 256 CBC, and stored encrypted. We strongly recommend to not use accounts with any value.
            </div>
            <div class="v-input mt-2 v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <fieldset aria-hidden="true">
                    <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                  </fieldset>
                  <div class="v-text-field__slot"><label for="privateKey" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Key*</label><input required="required" id="privateKey" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary" id="unlockAccount"><span class="v-btn__content">Unlock</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`UnlockAccountModal.vue Should let you unlock with a private key 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 430px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Unlock Account</div>
          <div class="v-card__text">
            <!---->
            <div role="alert" class="v-alert v-sheet theme--dark success">
              <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-check-circle theme--dark"></i>
                <div class="v-alert__content">Account unlocked.</div>
              </div>
            </div>
            <div>
              Set private key for <b>0x123</b> in order to use it for methods call. If you've already set one in the past, it will override it.
            </div>
            <div>
              Private keys are encrypted server side with AES 256 CBC, and stored encrypted. We strongly recommend to not use accounts with any value.
            </div>
            <div class="v-input mt-2 v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <fieldset aria-hidden="true">
                    <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                  </fieldset>
                  <div class="v-text-field__slot"><label for="privateKey" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Key*</label><input required="required" id="privateKey" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary" id="unlockAccount"><span class="v-btn__content">Unlock</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
